File: src/main/java/com/insogroup/utils/IOC/IOC.java
package com.insogroup.utils.IOC;

import com.insogroup.utils.ClassExtractor.exceptions.ClassExtractException;
import com.insogroup.utils.ClassExtractor.ClassExtractor;
import com.insogroup.utils.IOC.exceptions.ResolveStrategyException;
import com.insogroup.utils.IOC.interfaces.IOCStrategy;
import com.insogroup.utils.MailBoxes.TransmissionMailBox;
import com.insogroup.utils.MailBoxes.interfaces.MailBox;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * IOC container using for resolving some dependencies
 */
public class IOC {

    protected static final Map<String, IOCStrategy> strategies = new HashMap<>();
    protected static final TransmissionMailBox<String> targetMailBox = new TransmissionMailBox<>();

    /**
     * Register dependency in system
     * @param key Using for ident strategy in the future
     * @param strategy Some strategy
     */
    public static void register(String key, IOCStrategy strategy) {
        synchronized (strategies) {
            strategies.put(key, strategy);
        }
        targetMailBox.receive("Strategy for key \"" + key + "\" with strategy \"" + strategy.toString() + "\" registered.\n");
    }

    public static void subscribe(MailBox<String>... mailBoxes) {
        targetMailBox.subscribe(mailBoxes);
    }

    public static void unsubscribe(MailBox<String>... mailBoxes) {
        targetMailBox.unsubscribe(mailBoxes);
    }

    public static void subscribe(List<MailBox<String>> mailBoxes) {
        for (MailBox<String> mailBox : mailBoxes) {
            targetMailBox.subscribe(mailBox);
        }
    }

    public static void unsubscribe(List<MailBox<String>> mailBoxes) {
        for (MailBox<String> mailBox : mailBoxes) {
            targetMailBox.unsubscribe(mailBox);
        }
    }

    /**
     * Call for get dependency
     * @param key Key of dependency
     * @param args Args for dependency
     * @param <T> Type of target dependency
     * @return Resolved dependency
     * @throws ResolveStrategyException Throw when strategy is not registered
     */
    public static <T> T resolve(String key, Object... args) throws ResolveStrategyException{
        try {
            IOCStrategy strategy = strategies.get(key);
            return (T) strategy.getInstance(args);
        } catch (NullPointerException e) {
            try {
                return ClassExtractor.extract(key, args);
            } catch (ClassExtractException e1) {
                throw new ResolveStrategyException("Can't find strategy by key or extract class: " + key, e);
            }
        }
    }
}

File: src/main/kotlin/com/insogroup/utils/IOC/IOC.kt
package com.insogroup.utils.IOC

import com.insogroup.utils.ClassExtractor.exceptions.ClassExtractException
import com.insogroup.utils.ClassExtractor.ClassExtractor
import com.insogroup.utils.IOC.exceptions.ResolveStrategyException
import com.insogroup.utils.IOC.interfaces.IOCStrategy

import java.util.HashMap

/**
 * IOC container using for resolving some dependencies
 */
class IOC {
    private val strategies: MutableMap<String, IOCStrategy> = HashMap()
    private val subscribers: MutableList<(String) -> Unit> = ArrayList() 

    /**
     * Register dependency in system
     * @param key Using for ident strategy in the future
     * *
     * @param strategy Some strategy
     */
    fun register(key: String, strategy: IOCStrategy) {
        synchronized(strategies) {
            strategies.put(key, strategy)
        }
        subscribers.forEach {
            it("Strategy for key \"$key\" with strategy \"$strategy\" registered.\n")
        }
    }

    fun subscribe(vararg mailBoxes: (String) -> Unit) {
        subscribers.addAll(mailBoxes)
    }

    fun unsubscribe(vararg mailBoxes: (String) -> Unit) {
        subscribers.removeAll(mailBoxes)
    }

    fun subscribe(mailBoxes: List<(String) -> Unit>) {
        subscribers.addAll(mailBoxes)
    }

    fun unsubscribe(mailBoxes: List<(String) -> Unit>) {
        subscribers.removeAll(mailBoxes)
    }

    /**
     * Call for get dependency
     * @param key Key of dependency
     * *
     * @param args Args for dependency
     * *
     * @param <T> Type of target dependency
     * *
     * @return Resolved dependency
     * *
     * @throws ResolveStrategyException Throw when strategy is not registered
    </T> */
    @Throws(ResolveStrategyException::class)
    fun <T: Any> resolve(key: String, vararg args: Any): T {
        try {
            val strategy = strategies[key]
            return strategy!!.getInstance(*args) as T
        } catch (e: NullPointerException) {
            try {
                return ClassExtractor.extract<T>(key, *args)
            } catch (e1: ClassExtractException) {
                throw ResolveStrategyException("Can't find strategy by key or extract class: $key", e)
            }

        }

    }
}
